interval.hh

```diff
@@ -88,22 +89,178 @@ public:
     template <typename Transformer>
     using transformed_type = typename std::remove_cv_t<std::remove_reference_t<std::invoke_result_t<Transformer, T>>>;
 private:
-    optional<bound> _start;
-    optional<bound> _end;
-    bool _singular;
+    bool _start_exists = false;
+    bool _start_inclusive = false;
+    bool _end_exists = false;
+    bool _end_inclusive = false;
+    bool _singular = false;
+    union {
+        T _start_value;
+#ifdef DEBUG
+        uint64_t _start_poison;
+#endif
+    };
+    union {
+        T _end_value;
+#ifdef DEBUG
+        uint64_t _end_poison;
+#endif
+    };
 public:
-    wrapping_interval(optional<bound> start, optional<bound> end, bool singular = false)
-        : _start(std::move(start))
-        , _singular(singular) {
+    wrapping_interval(optional<bound> start, optional<bound> end, bool singular = false) {
+#ifdef DEBUG
+        _start_poison = 0xDEADBEEFDEADBEEF;
+        _end_poison = 0xDEADBEEFDEADBEEF;
```

Comment:
I'm afraid I don't understand why you made such an effort to do this "poison" thing. It seems you don't ever assert that poison is still set to that value, so what's the point? Also, it seems your code has a post-condition that always when the flag is true, the value will NOT be set to this poison. So why do we even need this poison feature - just to check for bugs in this optional<>-like code itself?

Reply:
I'll document it. The idea is that after poisoning, referencing the value incorrectly (say via the destructor) will crash. Otherwise, if it's zero (after moving a unique_ptr away, used in debug mode), a bad destroy doesn't trap.